config/config.go

```diff
@@ -0,0 +1,144 @@
+package config
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/bivas/rivi/config/action"
+	"github.com/bivas/rivi/config/client"
+	"github.com/bivas/rivi/engine"
+	"github.com/bivas/rivi/util"
+	"github.com/bivas/rivi/util/log"
+	"github.com/spf13/viper"
+)
+
+type Configuration interface {
+	GetClientConfig() client.ClientConfig
+	GetRoleMembers(role ...string) []string
+	GetRoles() []string
+	GetRules() []engine.Rule
+	GetActionConfig(kind string) (action.ActionConfig, error)
+}
+
+var (
+	configSections = []string{"config", "roles", "rules"}
+	lc             = log.Get("config")
+)
+
+type config struct {
+	internal      map[string]*viper.Viper
+	clientConfig  client.ClientConfig
+	rules         []engine.Rule
+	roles         map[string][]string
+	rolesKeys     []string
+	actionConfigs map[string]action.ActionConfig
+}
+
+func (c *config) GetActionConfig(kind string) (action.ActionConfig, error) {
+	config, exists := c.actionConfigs[kind]
+	if !exists {
+		return nil, fmt.Errorf("No such action config %s", kind)
+	}
+	return config, nil
+}
+
+func (c *config) GetClientConfig() client.ClientConfig {
+	return c.clientConfig
+}
+
+func (c *config) GetRoleMembers(roles ...string) []string {
+	result := make([]string, 0)
+	for _, role := range roles {
+		if members, ok := c.roles[role]; ok {
+			result = append(result, members...)
+		}
+	}
+	set := util.StringSet{Transformer: strings.ToLower}
```

Comment:
IMO the right place to normalize the names is when you read them from configuration.

Reply:
Agree. BUT do to the nature of interaction with viper, it will make the code look much harder to understand.